test/Test/Pos/Update/PollSpec.hs

```diff
@@ -6,24 +6,149 @@ module Test.Pos.Update.PollSpec
+actionToMonad (SetAdoptedBV bv) = Poll.setAdoptedBV bv
+actionToMonad (SetLastConfirmedSV sv) = Poll.setLastConfirmedSV sv
+actionToMonad (DelConfirmedSV an) = Poll.delConfirmedSV an
+actionToMonad (AddConfirmedProposal cps) = Poll.addConfirmedProposal cps
+actionToMonad (DelConfirmedProposal sv) = Poll.delConfirmedProposal sv
+actionToMonad (InsertActiveProposal ps) = Poll.insertActiveProposal ps
+actionToMonad (DeactivateProposal ui) = Poll.deactivateProposal ui
+actionToMonad (SetSlottingData sd) = Poll.setSlottingData sd
+actionToMonad (SetEpochProposers hs) = Poll.setEpochProposers hs
+
+applyActionToModifier :: PollAction -> Poll.PollModifier -> Poll.PollModifier
+applyActionToModifier (PutBVState bv bvs) = Poll.pmBVsL %~ MM.insert bv bvs
+applyActionToModifier (DelBVState bv) = Poll.pmBVsL %~ MM.delete bv
+applyActionToModifier (SetAdoptedBV bv) = \p ->
+    case MM.lookup (const Nothing) bv (Poll.pmBVs p) of
+        Nothing -> p
+        Just (Poll.bvsData -> bvd) -> p { Poll.pmAdoptedBVFull = Just (bv, bvd) }
+applyActionToModifier (SetLastConfirmedSV SoftwareVersion {..}) =
+    Poll.pmConfirmedL %~ MM.insert svAppName svNumber
+applyActionToModifier (DelConfirmedSV an) = Poll.pmConfirmedL %~ MM.delete an
+applyActionToModifier (AddConfirmedProposal cps) =
+    Poll.pmConfirmedPropsL %~ MM.insert (Poll.cpsSoftwareVersion cps) cps
+applyActionToModifier (DelConfirmedProposal sv) = Poll.pmConfirmedPropsL %~ MM.delete sv
+applyActionToModifier (InsertActiveProposal ps) = \p ->
+    let up@Poll.UnsafeUpdateProposal{upSoftwareVersion = sv, ..} = Poll.psProposal ps
+        upId = hash up
+        appName = svAppName sv
+        alterDel _ Nothing     = Nothing
+        alterDel val (Just hs) = Just $ HS.delete val hs
+        p' = case MM.lookup (const Nothing) upId (Poll.pmActiveProps p) of
+            Nothing -> p
+            Just _ -> p & Poll.pmEpochProposersL %~ fmap (HS.insert (addressHash upFrom))
+    in p' & ((Poll.pmActivePropsL %~ MM.insert upId ps) .
+            (Poll.pmDelActivePropsIdxL %~ HM.alter (alterDel upId) appName))
+applyActionToModifier (DeactivateProposal ui) = \p ->
+    let proposal = MM.lookup (const Nothing) ui (Poll.pmActiveProps p)
+    in case proposal of
+           Nothing -> p
+           Just ps ->
+               let up = Poll.psProposal ps
+                   upId = hash up
+                   sv = Poll.upSoftwareVersion up
+                   appName = svAppName sv
+                   alterIns val Nothing   = Just $ HS.singleton val
+                   alterIns val (Just hs) = Just $ HS.insert val hs
+               in p & ((Poll.pmActivePropsL %~ MM.delete upId) .
+                        (Poll.pmDelActivePropsIdxL %~ HM.alter (alterIns upId) appName))
+
+applyActionToModifier (SetSlottingData sd) = Poll.pmSlottingDataL .~ (Just sd)
+applyActionToModifier (SetEpochProposers hs) = Poll.pmEpochProposersL .~ (Just hs)
+
+instance Arbitrary PollAction where
+    arbitrary = oneof
```

Comment:
@int-index I had originally written a more complex `Arbitrary` instance but ended up changing it to this one. I should've replaced with this, you're right.